src/cdp/Node.zig

```diff
@@ -201,49 +201,69 @@ pub const Search = struct {
 // (For now, we only support direct children)
 
 pub const Writer = struct {
-    opts: Opts,
-    node: *const Node,
+    depth: i32,
+    exclude_root: bool,
+    root: *const Node,
     registry: *Registry,
 
-    pub const Opts = struct {};
+    pub const Opts = struct {
+        depth: i32 = 0,
+        exclude_root: bool = false,
+    };
 
     pub fn jsonStringify(self: *const Writer, w: anytype) !void {
-        self.toJSON(w) catch |err| {
-            // The only error our jsonStringify method can return is
-            // @TypeOf(w).Error. In other words, our code can't return its own
-            // error, we can only return a writer error. Kinda sucks.
-            log.err(.cdp, "json stringify", .{ .err = err });
-            return error.OutOfMemory;
-        };
+        if (self.exclude_root) {
+            _ = self.writeChildren(self.root, 0, w) catch |err| {
+                log.err(.cdp, "node writeChildren", .{ .err = err });
+                return error.OutOfMemory;
+            };
+        } else {
+            self.toJSON(self.root, 0, w) catch |err| {
+                // The only error our jsonStringify method can return is
+                // @TypeOf(w).Error. In other words, our code can't return its own
+                // error, we can only return a writer error. Kinda sucks.
+                log.err(.cdp, "node toJSON stringify", .{ .err = err });
+                return error.OutOfMemory;
+            };
+        }
     }
 
-    fn toJSON(self: *const Writer, w: anytype) !void {
+    fn toJSON(self: *const Writer, node: *const Node, depth: usize, w: anytype) !void {
         try w.beginObject();
-        try self.writeCommon(self.node, false, w);
+        try self.writeCommon(node, false, w);
 
-        {
-            var registry = self.registry;
-            const child_nodes = try parser.nodeGetChildNodes(self.node._node);
-            const child_count = try parser.nodeListLength(child_nodes);
+        try w.objectField("children");
+        const child_count = try self.writeChildren(node, depth, w);
+        try w.objectField("childNodeCount");
+        try w.write(child_count);
 
-            var i: usize = 0;
-            try w.objectField("children");
-            try w.beginArray();
-            for (0..child_count) |_| {
-                const child = (try parser.nodeListItem(child_nodes, @intCast(i))) orelse break;
-                const child_node = try registry.register(child);
+        try w.endObject();
+    }
+
+    fn writeChildren(self: *const Writer, node: *const Node, depth: usize, w: anytype) anyerror!usize {
+        var registry = self.registry;
+        const child_nodes = try parser.nodeGetChildNodes(node._node);
+        const child_count = try parser.nodeListLength(child_nodes);
+        const full_child = self.depth < 0 or self.depth < depth;
```

Comment:
```suggestion
        const full_child = self.depth < 0 or depth < self.depth;
```
My understanding is that if  `self.depth` is 1 we only get one layer of nodes.
So in the `exclude_root == true` case that would be a single list of nodes without any of their children populated.
So I would expect this function to be called with the parameter `depth: usize` set to `1` (now called with `0`).

In the `exclude_root == false` I would expect `writeChildren` not to be called at all as just the root node would be written. So we'd need this if statement also in the `toJSON` function.

I would not be surpised if I'm off by one tho.